File: julia/AIInAction/src/ch131_spectral_clustering.jl
"""
    Ch131SpectralClustering

Spectral clustering from scratch (chapter 131). Mirrors the Python module
`aiinaction.ch131_spectral_clustering` and the Rust module
`aiinaction::ch131_spectral_clustering` one-to-one. The eigensolver (cyclic
Jacobi) and k-means are implemented from scratch and made deterministic so all
three language ports agree to within `1e-9` on the shared fixtures.
"""
module Ch131SpectralClustering

export rbf_affinity, normalized_laplacian, jacobi_eigh,
    spectral_embedding, kmeans, spectral_clustering

function _validate_matrix(X::AbstractMatrix{<:Real})
    size(X, 1) == 0 && throw(ArgumentError("X must contain at least one sample"))
    all(isfinite, X) || throw(ArgumentError("X must contain only finite values"))
    return nothing
end

"""Gaussian (RBF) similarity matrix with a zero diagonal."""
function rbf_affinity(X::AbstractMatrix{<:Real}, sigma::Real)
    _validate_matrix(X)
    (isfinite(sigma) && sigma > 0) ||
        throw(ArgumentError("sigma must be a positive finite number; got $sigma"))
    n = size(X, 1)
    denom = 2.0 * sigma * sigma
    W = zeros(Float64, n, n)
    for i in 1:n
        for j in (i + 1):n
            diff = @view(X[i, :]) .- @view(X[j, :])
            w = exp(-sum(abs2, diff) / denom)
            W[i, j] = w
            W[j, i] = w
        end
    end
    return W
end

"""Symmetric normalized Laplacian `L_sym = I - D^{-1/2} W D^{-1/2}`."""
function normalized_laplacian(W::AbstractMatrix{<:Real})
    n = size(W, 1)
    n == 0 && throw(ArgumentError("W must be non-empty"))
    size(W, 2) == n || throw(ArgumentError("W must be a square matrix"))
    all(isfinite, W) || throw(ArgumentError("W must contain only finite values"))
    all(>=(0), W) || throw(ArgumentError("W must have non-negative entries"))
    for i in 1:n, j in 1:n
        abs(W[i, j] - W[j, i]) > 1e-12 && throw(ArgumentError("W must be symmetric"))
    end
    deg = vec(sum(W, dims=2))
    all(>(0), deg) ||
        throw(ArgumentError("every vertex must have positive degree (no isolated vertices)"))
    dinv = 1.0 ./ sqrt.(deg)
    L = Matrix{Float64}(undef, n, n)
    for i in 1:n, j in 1:n
        id = i == j ? 1.0 : 0.0
        L[i, j] = id - dinv[i] * W[i, j] * dinv[j]
    end
    for i in 1:n, j in (i + 1):n
        m = 0.5 * (L[i, j] + L[j, i])
        L[i, j] = m
        L[j, i] = m
    end
    return L
end

"""
Eigen-decomposition of a real symmetric matrix via cyclic Jacobi rotations.
Returns `(eigenvalues, eigenvectors)` with ascending eigenvalues and
sign-fixed eigenvectors as columns.
"""
function jacobi_eigh(A::AbstractMatrix{<:Real})
    n = size(A, 1)
    n == 0 && throw(ArgumentError("A must be non-empty"))
    size(A, 2) == n || throw(ArgumentError("A must be a square matrix"))
    for i in 1:n, j in 1:n
        abs(A[i, j] - A[j, i]) > 1e-12 && throw(ArgumentError("A must be symmetric"))
    end
    max_sweeps = 100
    tol = 1e-12
    M = Matrix{Float64}(A)
    V = Matrix{Float64}(I_eye(n))

    for _ in 1:max_sweeps
        off = 0.0
        for p in 1:n, q in (p + 1):n
            off += M[p, q]^2
        end
        sqrt(off) <= tol && break
        for p in 1:n
            for q in (p + 1):n
                apq = M[p, q]
                abs(apq) <= 1e-300 && continue
                app = M[p, p]
                aqq = M[q, q]
                phi = 0.5 * atan(2.0 * apq, aqq - app)
                c = cos(phi)
                s = sin(phi)
                for i in 1:n
                    mip = M[i, p]
                    miq = M[i, q]
                    M[i, p] = c * mip - s * miq
                    M[i, q] = s * mip + c * miq
                end
                for i in 1:n
                    mpi = M[p, i]
                    mqi = M[q, i]
                    M[p, i] = c * mpi - s * mqi
                    M[q, i] = s * mpi + c * mqi
                end
                for i in 1:n
                    vip = V[i, p]
                    viq = V[i, q]
                    V[i, p] = c * vip - s * viq
                    V[i, q] = s * vip + c * viq
                end
            end
        end
    end

    eig = [M[i, i] for i in 1:n]
    order = sortperm(eig)
    eigvals = eig[order]
    vecs = V[:, order]
    for j in 1:n
        for i in 1:n
            if abs(vecs[i, j]) > 1e-12
                if vecs[i, j] < 0.0
                    @views vecs[:, j] .= .-vecs[:, j]
                end
                break
            end
        end
    end
    return eigvals, vecs
end

# Small identity helper to avoid a LinearAlgebra dependency.
function I_eye(n::Int)
    M = zeros(Float64, n, n)
    for i in 1:n
        M[i, i] = 1.0
    end
    return M
end

"""Row-normalized spectral embedding: the `k` smallest eigenvectors of `L_sym`."""
function spectral_embedding(W::AbstractMatrix{<:Real}, k::Integer)
    L = normalized_laplacian(W)
    n = size(L, 1)
    (1 <= k <= n) || throw(ArgumentError("k must be an integer in [1, n]=$n; got $k"))
    _, vecs = jacobi_eigh(L)
    U = Matrix{Float64}(vecs[:, 1:k])
    for i in 1:n
        nrm = sqrt(sum(abs2, @view U[i, :]))
        if nrm > 1e-12
            @views U[i, :] ./= nrm
        end
    end
    return U
end

"""Lloyd's k-means with deterministic furthest-point seeding. Returns `(labels, centers)`."""
function kmeans(X::AbstractMatrix{<:Real}, k::Integer)
    _validate_matrix(X)
    n, d = size(X)
    (1 <= k <= n) || throw(ArgumentError("k must be an integer in [1, n]=$n; got $k"))
    max_iter = 100
    tol = 1e-10

    first = 1
    for i in 2:n
        if _lex_less(@view(X[i, :]), @view(X[first, :]))
            first = i
        end
    end
    center_idx = [first]
    while length(center_idx) < k
        best_i = -1
        best_d = -1.0
        for i in 1:n
            i in center_idx && continue
            dist = minimum(sum(abs2, @view(X[i, :]) .- @view(X[c, :])) for c in center_idx)
            if dist > best_d
                best_d = dist
                best_i = i
            end
        end
        push!(center_idx, best_i)
    end
    centers = Matrix{Float64}(X[center_idx, :])

    labels = zeros(Int, n)
    for _ in 1:max_iter
        for i in 1:n
            best_c = 1
            best_d = Inf
            for c in 1:k
                dist = sum(abs2, @view(X[i, :]) .- @view(centers[c, :]))
                if dist < best_d
                    best_d = dist
                    best_c = c
                end
            end
            labels[i] = best_c
        end
        new_centers = copy(centers)
        for c in 1:k
            members = [i for i in 1:n if labels[i] == c]
            if !isempty(members)
                for t in 1:d
                    new_centers[c, t] = sum(X[i, t] for i in members) / length(members)
                end
            end
        end
        shift = sqrt(sum(abs2, new_centers .- centers))
        centers = new_centers
        shift <= tol && break
    end
    # Return labels in 0-based form to match the Python/Rust ports.
    return labels .- 1, centers
end

function _lex_less(a, b)
    for (x, y) in zip(a, b)
        x < y && return true
        x > y && return false
    end
    return false
end

"""End-to-end spectral clustering (Ng-Jordan-Weiss). Returns 0-based labels."""
function spectral_clustering(X::AbstractMatrix{<:Real}, k::Integer, sigma::Real=1.0)
    W = rbf_affinity(X, sigma)
    embedding = spectral_embedding(W, k)
    labels, _ = kmeans(embedding, k)
    return labels
end

end # module Ch131SpectralClustering

File: src/aiinaction/ch131_spectral_clustering.py
"""Spectral clustering from scratch (chapter 131).

A small, dependency-free reference implementation of spectral clustering built
directly on the normalized-cut relaxation developed in the chapter. The pipeline
is the classic one:

1. Build a Gaussian (RBF) similarity matrix ``W`` from the data.
2. Form the degree matrix ``D`` and the symmetric normalized Laplacian
   ``L_sym = I - D^{-1/2} W D^{-1/2}``.
3. Take the ``k`` eigenvectors of ``L_sym`` with the smallest eigenvalues.
4. Row-normalize that ``n x k`` embedding (the Ng-Jordan-Weiss step).
5. Cluster the embedded rows with k-means.

Every numerical primitive here, including the eigensolver (a cyclic Jacobi
rotation sweep for symmetric matrices) and k-means, is implemented from scratch
and made fully deterministic so that the Python, Julia, and Rust ports agree to
within ``1e-9`` on the shared fixtures. ``numpy`` is used only as a convenient
array container and for elementary linear-algebra products; no high-level
clustering or eigen routine is called.

This module mirrors the Julia (``AIInAction.Ch131SpectralClustering``) and Rust
(``aiinaction::ch131_spectral_clustering``) implementations one-to-one.
"""
from __future__ import annotations

import math
from collections.abc import Sequence

import numpy as np

__all__ = [
    "rbf_affinity",
    "normalized_laplacian",
    "jacobi_eigh",
    "spectral_embedding",
    "kmeans",
    "spectral_clustering",
]


def _as_matrix(x: Sequence[Sequence[float]]) -> np.ndarray:
    """Coerce a 2-D sequence into a float ``(n, d)`` array with validation."""
    arr = np.asarray(x, dtype=float)
    if arr.ndim != 2:
        raise ValueError(f"X must be a 2-D array of shape (n_samples, n_features); got ndim={arr.ndim}")
    if arr.shape[0] == 0:
        raise ValueError("X must contain at least one sample")
    if not np.all(np.isfinite(arr)):
        raise ValueError("X must contain only finite values")
    return arr


def rbf_affinity(X: Sequence[Sequence[float]], sigma: float) -> np.ndarray:
    """Gaussian (RBF) similarity matrix with a zero diagonal.

    The off-diagonal entries are ``w_ij = exp(-||x_i - x_j||^2 / (2 sigma^2))``
    and ``w_ii = 0`` so that vertices carry no self-loop, matching the chapter's
    convention ``w_ii = 0``.

    Args:
        X: Data of shape ``(n_samples, n_features)``.
        sigma: Positive kernel bandwidth.

    Returns:
        A symmetric ``(n, n)`` affinity matrix with zero diagonal.

    Raises:
        ValueError: If ``X`` is malformed or ``sigma`` is not positive.
    """
    arr = _as_matrix(X)
    if not math.isfinite(sigma) or sigma <= 0.0:
        raise ValueError(f"sigma must be a positive finite number; got {sigma}")
    n = arr.shape[0]
    denom = 2.0 * sigma * sigma
    W = np.zeros((n, n), dtype=float)
    for i in range(n):
        for j in range(i + 1, n):
            diff = arr[i] - arr[j]
            sq = float(diff @ diff)
            w = math.exp(-sq / denom)
            W[i, j] = w
            W[j, i] = w
    return W


def normalized_laplacian(W: Sequence[Sequence[float]]) -> np.ndarray:
    """Symmetric normalized Laplacian ``L_sym = I - D^{-1/2} W D^{-1/2}``.

    Args:
        W: A square symmetric non-negative affinity matrix.

    Returns:
        The symmetric normalized Laplacian as an ``(n, n)`` array.

    Raises:
        ValueError: If ``W`` is not square, not symmetric, has negative
            entries, or has an isolated vertex (zero degree).
    """
    A = np.asarray(W, dtype=float)
    if A.ndim != 2 or A.shape[0] != A.shape[1]:
        raise ValueError("W must be a square 2-D matrix")
    n = A.shape[0]
    if n == 0:
        raise ValueError("W must be non-empty")
    if not np.all(np.isfinite(A)):
        raise ValueError("W must contain only finite values")
    if np.any(A < 0.0):
        raise ValueError("W must have non-negative entries")
    if not np.allclose(A, A.T, atol=1e-12, rtol=0.0):
        raise ValueError("W must be symmetric")
    deg = A.sum(axis=1)
    if np.any(deg <= 0.0):
        raise ValueError("every vertex must have positive degree (no isolated vertices)")
    dinv_sqrt = 1.0 / np.sqrt(deg)
    L = np.eye(n) - (dinv_sqrt[:, None] * A * dinv_sqrt[None, :])
    # Symmetrize to remove any floating-point asymmetry.
    return 0.5 * (L + L.T)


def jacobi_eigh(A: Sequence[Sequence[float]], max_sweeps: int = 100, tol: float = 1e-12) -> tuple[np.ndarray, np.ndarray]:
    """Eigen-decomposition of a real symmetric matrix via cyclic Jacobi rotations.

    This is a deterministic, dependency-free eigensolver (no ``numpy.linalg``)
    so that all three language ports produce bitwise-comparable results. It
    returns eigenvalues in ascending order together with the matching
    orthonormal eigenvectors as columns, with each eigenvector sign-fixed so its
    first nonzero entry is positive (removing the inherent sign ambiguity).

    Args:
        A: A square symmetric matrix.
        max_sweeps: Maximum number of full off-diagonal sweeps.
        tol: Convergence threshold on the off-diagonal Frobenius norm.

    Returns:
        Tuple ``(eigenvalues, eigenvectors)`` with eigenvalues ascending and
        eigenvectors as columns of an orthogonal matrix.

    Raises:
        ValueError: If ``A`` is not square or not symmetric.
    """
    M = np.array(A, dtype=float)
    if M.ndim != 2 or M.shape[0] != M.shape[1]:
        raise ValueError("A must be a square 2-D matrix")
    n = M.shape[0]
    if n == 0:
        raise ValueError("A must be non-empty")
    if not np.allclose(M, M.T, atol=1e-12, rtol=0.0):
        raise ValueError("A must be symmetric")

    V = np.eye(n)
    for _ in range(max_sweeps):
        off = math.sqrt(sum(M[p, q] ** 2 for p in range(n) for q in range(p + 1, n)))
        if off <= tol:
            break
        for p in range(n):
            for q in range(p + 1, n):
                apq = M[p, q]
                if abs(apq) <= 1e-300:
                    continue
                app = M[p, p]
                aqq = M[q, q]
                # Jacobi rotation angle that zeroes M[p, q].
                phi = 0.5 * math.atan2(2.0 * apq, aqq - app)
                c = math.cos(phi)
                s = math.sin(phi)
                for i in range(n):
                    mip = M[i, p]
                    miq = M[i, q]
                    M[i, p] = c * mip - s * miq
                    M[i, q] = s * mip + c * miq
                for i in range(n):
                    mpi = M[p, i]
                    mqi = M[q, i]
                    M[p, i] = c * mpi - s * mqi
                    M[q, i] = s * mpi + c * mqi
                for i in range(n):
                    vip = V[i, p]
                    viq = V[i, q]
                    V[i, p] = c * vip - s * viq
                    V[i, q] = s * vip + c * viq

    eigvals = np.array([M[i, i] for i in range(n)], dtype=float)
    order = sorted(range(n), key=lambda i: eigvals[i])
    eigvals = eigvals[order]
    V = V[:, order]
    # Sign-fix each eigenvector: first nonzero entry positive.
    for j in range(n):
        for i in range(n):
            if abs(V[i, j]) > 1e-12:
                if V[i, j] < 0.0:
                    V[:, j] = -V[:, j]
                break
    return eigvals, V


def spectral_embedding(W: Sequence[Sequence[float]], k: int) -> np.ndarray:
    """Row-normalized spectral embedding from the normalized Laplacian.

    Builds ``L_sym``, takes the ``k`` smallest eigenvectors, and normalizes each
    row to unit length (the Ng-Jordan-Weiss embedding). Rows that are exactly
    zero are left as zero.

    Args:
        W: Affinity matrix of shape ``(n, n)``.
        k: Number of eigenvectors / target clusters, ``1 <= k <= n``.

    Returns:
        An ``(n, k)`` embedding whose rows are clustered by k-means downstream.

    Raises:
        ValueError: If ``k`` is out of range.
    """
    L = normalized_laplacian(W)
    n = L.shape[0]
    if not isinstance(k, int) or k < 1 or k > n:
        raise ValueError(f"k must be an integer in [1, n]={n}; got {k}")
    _, vecs = jacobi_eigh(L)
    U = vecs[:, :k].copy()
    norms = np.sqrt((U * U).sum(axis=1))
    nonzero = norms > 1e-12
    U[nonzero] = U[nonzero] / norms[nonzero, None]
    return U


def kmeans(
    X: Sequence[Sequence[float]],
    k: int,
    max_iter: int = 100,
    tol: float = 1e-10,
) -> tuple[list[int], np.ndarray]:
    """Lloyd's k-means with deterministic furthest-point seeding.

    The initialization is fully deterministic: the first center is the data row
    with the smallest (lexicographic) coordinates, and each subsequent center is
    the point furthest from the set of chosen centers, ties broken by lowest
    index. This removes the random seeding that would otherwise prevent
    cross-language parity.

    Args:
        X: Points of shape ``(n_samples, n_features)``.
        k: Number of clusters, ``1 <= k <= n``.
        max_iter: Maximum Lloyd iterations.
        tol: Convergence threshold on the total center shift.

    Returns:
        Tuple ``(labels, centers)`` where ``labels[i]`` is the cluster index of
        point ``i`` and ``centers`` has shape ``(k, n_features)``.

    Raises:
        ValueError: If ``X`` is malformed or ``k`` is out of range.
    """
    arr = _as_matrix(X)
    n = arr.shape[0]
    if not isinstance(k, int) or k < 1 or k > n:
        raise ValueError(f"k must be an integer in [1, n]={n}; got {k}")

    # Deterministic seed 0: lexicographically smallest point.
    first = min(range(n), key=lambda i: tuple(arr[i]))
    center_idx = [first]
    while len(center_idx) < k:
        best_i = -1
        best_d = -1.0
        for i in range(n):
            if i in center_idx:
                continue
            d = min(float((arr[i] - arr[c]) @ (arr[i] - arr[c])) for c in center_idx)
            if d > best_d:
                best_d = d
                best_i = i
        center_idx.append(best_i)
    centers = arr[center_idx].astype(float).copy()

    labels = [0] * n
    for _ in range(max_iter):
        # Assignment step (ties broken by lowest cluster index).
        for i in range(n):
            best_c = 0
            best_d = float("inf")
            for c in range(k):
                diff = arr[i] - centers[c]
                d = float(diff @ diff)
                if d < best_d:
                    best_d = d
                    best_c = c
            labels[i] = best_c
        # Update step. Empty clusters keep their previous center.
        new_centers = centers.copy()
        for c in range(k):
            members = [i for i in range(n) if labels[i] == c]
            if members:
                new_centers[c] = arr[members].mean(axis=0)
        shift = float(np.sqrt(((new_centers - centers) ** 2).sum()))
        centers = new_centers
        if shift <= tol:
            break
    return labels, centers


def spectral_clustering(
    X: Sequence[Sequence[float]],
    k: int,
    sigma: float = 1.0,
) -> list[int]:
    """End-to-end spectral clustering (Ng-Jordan-Weiss).

    Builds an RBF affinity matrix, computes the row-normalized spectral
    embedding from the symmetric normalized Laplacian, and clusters the rows
    with deterministic k-means.

    Args:
        X: Data of shape ``(n_samples, n_features)``.
        k: Number of clusters, ``1 <= k <= n_samples``.
        sigma: Positive RBF bandwidth.

    Returns:
        A list of integer cluster labels, one per sample.

    Raises:
        ValueError: If any argument is malformed.
    """
    W = rbf_affinity(X, sigma)
    embedding = spectral_embedding(W, k)
    labels, _ = kmeans(embedding, k)
    return labels
